File: C-implementation/src/lacoperon.c
#include "all.h"
#include "lacoperon.h"

//   equation(3)
double dMRdt(double* params, double* last) {
    return params[ksMR]
           -(params[lMR]*last[MR]);
}

//   equation(4)
double dRdt(double* params, double* last) {
    return (params[ksR]*last[MR])
           -(2*params[k2R]*(last[R]*last[R]))
           +(2*params[kn2R]*last[R2])
           -(params[lR]*last[R]);
}

//   equation(5)
double dR2dt(double* params, double* last) {
    return (params[k2R]*(last[R]*last[R]))
           -(params[kn2R]*last[R2])
           -(params[kr]*last[R2]*last[O])
           +(params[knr]*(params[OT]-last[O]))
           -(params[kdr1]*last[R2]*(last[I]*last[I]))
           +(params[kndr1]*last[I2R2])
           -(params[lR2]*last[R2]);
}

//   equation(6)
double dOdt(double* params, double* last) {
    return -(params[kr]*last[R2]*last[O])
           +(params[knr]*(params[OT]-last[O]))
           +(params[kdr2]*(params[OT]-last[O])*(last[I]*last[I]))
           -(params[kndr2]*last[O]*last[I2R2]);
}

//   equation(7)
double dIdt(double* params, double* last) {
    return -(2*params[kdr1]*last[R2]*(last[I]*last[I]))
           +(2*params[kndr1]*last[I2R2])
           -(2*params[kdr2]*(params[OT]-last[O])*(last[I]*last[I]))
           +(2*params[kndr2]*last[O]*last[I2R2])
           +(params[kft]*last[YIex])
           +(params[kt]*(params[Iex]-last[I]))
           +(2*params[lI2R2]*last[I2R2])
           +(params[lYIex]*last[YIex]);
}
//   equation(8)
double dI2R2dt(double* params, double* last) {
    return (params[kdr1]*last[R2]*(last[I]*last[I]))
           -(params[kndr1]*last[I2R2])
           +(params[kdr2]*(params[OT]-last[O])*(last[I]*last[I]))
           -(params[kndr2]*last[O]*last[I2R2])
           -(params[lI2R2]*last[I2R2]);
}

//   equation(9)
double dMYdt(double* params, double* last) {
    return (params[ks0MY]*(params[OT]-last[O]))
           +(params[ks1MY]*last[O])
           -(params[lMY]*last[MY]);
}

//   equation(10)
double dYdt(double* params, double* last) {
    return (params[ksY]*last[MY])
           +((params[kft]+params[knp])*last[YIex])
           -(params[kp]*last[Y]*params[Iex])
           -(params[lY]*last[Y]);
}
//   equation(11)H
double dYIexdt(double* params, double* last) {
    return -((params[kft]+params[knp])*last[YIex])
           +(params[kp]*last[Y]*params[Iex])
           -(params[lYIex]*last[YIex]);
}

Model lacoperon;
ode funcs[9] = {&dMRdt, &dRdt, &dR2dt, &dOdt, &dIdt, &dI2R2dt, &dMYdt, &dYdt, &dYIexdt};

int getNumSpecies() {
  return nS;
}

int getNumParams() {
  return nP;
}

const char** getNames() {
  return names;
}

Model getModel(double* initVals, double* params) {
    if (initVals) {
        lacoperon.initValues = initVals;
    } else {
        lacoperon.initValues = definitValues;
    }

    if (params) {
        lacoperon.params = params;
    } else {
        lacoperon.params = defparams;
    }

    lacoperon.funcs = funcs;
    lacoperon.numSpecies = nS;

    return lacoperon;
}

File: C-implementation/LacOperon.py
from lib import *
from util import *
#indices
MR = 0
R = 1
R2 = 2
O = 3
I = 4
I2R2 = 5
MY = 6
Y = 7
YIex = 8

hmlabels = ['MR dPK', 'MR dHFT', 'R dPK', 'R dPKT', 'R dHRT', 'R dHFT', 'R2 dPK', 'R2 dPKT', 'R2 dHRT', 'R2 dHFT', 'O dPK', 'O dPKT', 'O dHRT', 'O dHFT', 'I dPK', 'I dPKT', 'I dHRT', 'I dHFT', 'I2R2 dPK', 'I2R2 dPKT', 'I2R2 dHRT', 'I2R2 dHFT', 'MY dPK', 'MY dPKT', 'MY dHRT', 'MY dHFT', 'Y dPK', 'Y dPKT', 'Y dHRT', 'Y dHFT', 'YIex dPK', 'YIex dPKT', 'YIex dHRT', 'YIex dHFT']


param_names = ["OT","ksMR","ksR","k2R","kn2R","kr","knr","kdr1","kndr1","kdr2",
"kndr2","ks1MY","ks0MY","ksY","kp","knp","kft","kt","lMR","lMY",
"lR","lR2","lY","lYIex","lI2R2","Iex"]

def makeSensitivityList(se):
    onePls = list()
    for i in xrange(len(se)):
        seSp = se[i]
        if i == MR:
            onePls += [seSp[0][0]] + [seSp[1][2]]
        else:
            onePls += [seSp[0][0]] + seSp[1]
    return onePls

def makeSignChanges(matrix):
    matrix[6] = [-1*x for x in matrix[6]]
    return matrix


def lacOpCharFunc(data, boluspoint):
    startind = boluspoint - 1
    partial = 0.5
    vals = data.vals
    time = data.time
    size = data.size

    mx = lambda p,n: p > 0 and n < 0
    mn = lambda p,n: p < 0 and n > 0

    out = []
    for x in xrange(9):
        if x in [I, Y, YIex, MY, O]:
            # print("ind:",x)
            out.append(package(vals[x], time, find2HalfsPeak(vals[x], size, mn, startind, partial)))
        else:
            out.append(package(vals[x], time, find2HalfsPeak(vals[x], size, mx, startind, partial)))
    return out

class LacOperon(Model):
    """docstring for LacOperon"""
    def __init__(self):
        super(LacOperon, self).__init__("lacoperon")
        self.charFunc = lacOpCharFunc